Jetsurvey/app/src/main/java/com/example/compose/jetsurvey/survey/question/components/Draggable.kt

```diff
@@ -0,0 +1,187 @@
+                            translationY = offset.y.minus(targetSize.center.y)
+                        }
+                ) {
+                    state.draggableComposable?.invoke(false)
+                }
+            }
+        }
+    }
+}
+
+@Composable
+fun <T> DragTarget(
+    onItemDropped: (T) -> Unit,
+    item: T,
+    modifier: Modifier = Modifier,
+    content: @Composable ((isDragging: Boolean) -> Unit)
+) {
+    var currentPosition by remember { mutableStateOf(Offset.Zero) }
+    val state = LocalDragTargetInfo.current
+
+    // Local dragging states
+    var isItemDragging by remember { mutableStateOf(false) }
+    var hasItemDraggingEnd by remember { mutableStateOf(false) }
+
+    // Check if any item dropped and this is the item
+    if (state.hasDragEndInAnyDropZone && hasItemDraggingEnd) {
+        hasItemDraggingEnd = false
+        state.hasDragEndInAnyDropZone = false
+        onItemDropped(item)
+    }
+
+    if (isItemDragging) {
+        state.dataToTransfer = item
+    }
+
+    Box(
+        modifier = modifier
+            .onGloballyPositioned {
+                currentPosition = it.positionInWindow()
+            }
+            .pointerInput(Unit) {
+                detectDragGestures(
+                    onDragStart = {
+                        isItemDragging = true
+                        state.isDragging = true
+                        state.dragPosition = currentPosition + it
+                        state.draggableComposable = content
+                        state.hasDragEnd = false
+                    }, onDrag = { change, dragAmount ->
+                        change.consume()
+                        state.dragOffset += Offset(dragAmount.x, dragAmount.y)
+                    }, onDragEnd = {
+                        state.hasDragEnd = true
+                        state.isDragging = false
+                        state.dragOffset = Offset.Zero
+                        isItemDragging = false
+                        hasItemDraggingEnd = true
+                    }, onDragCancel = {
+                        state.dragOffset = Offset.Zero
+                        state.isDragging = false
+                        isItemDragging = false
+                    }
+                )
+            }
+    ) {
+        content(isItemDragging)
+    }
+}
+
+@Composable
+fun DropZone(
+    modifier: Modifier = Modifier,
+    isDropZone: Boolean = true,
+    content: @Composable BoxScope.(
+        isDraggingInDropZone: Boolean,
+        hasDragEndInDropZone: Boolean
+    ) -> Unit
+) {
+    val state = LocalDragTargetInfo.current
+    val dragPosition = state.dragPosition
+    val dragOffset = state.dragOffset
+    var isItemInTheDropZone by remember { mutableStateOf(false) }
+
+    val hasDragEndInDropZone by remember {
+        derivedStateOf { isItemInTheDropZone && state.hasDragEnd }
+    }
+    if (hasDragEndInDropZone) {
+        state.hasDragEndInAnyDropZone = true
```

Comment:
Our advice is to never write to state in a composition as it can cause unintended side effects https://developer.android.com/jetpack/compose/performance/bestpractices#avoid-backwards